services/github/github-total-star.service.js

```diff
@@ -0,0 +1,248 @@
+ */
+const validateAffiliations = (value, helpers) => {
+  const values = value.split(',')
+  if (values.some(e => !affiliationsAllowedValues.includes(e))) {
+    return helpers.error('any.invalid')
+  }
+  return value
+}
+
+const queryParamSchema = Joi.object({
+  org: Joi.valid('').optional(),
+  affiliations: Joi.string().default('OWNER').custom(validateAffiliations),
+}).required()
+
+module.exports = class GithubTotalStarService extends GithubAuthV4Service {
+  static defaultLabel = 'Stars'
+  static category = 'social'
+
+  static route = {
+    base: 'github/stars',
+    pattern: ':user',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: "Github User's stars",
+      namedParams: {
+        user: 'chris48s',
+      },
+      queryParams: { affiliations: 'OWNER,COLLABORATOR' },
+      staticPreview: {
+        label: this.defaultLabel,
+        message: 54,
+        style: 'social',
+      },
+      documentation: userDocumentation,
+    },
+    {
+      title: "Github Org's stars",
+      namedParams: {
+        user: 'badges',
+      },
+      queryParams: { org: null },
+      staticPreview: {
+        label: this.defaultLabel,
+        message: metric(7000),
+        style: 'social',
+      },
+      documentation: orgDocumentation,
+    },
+  ]
+
+  static defaultBadgeData = {
+    label: this.defaultLabel,
+    namedLogo: 'github',
+  }
+
+  static render({ totalStars, user }) {
+    return {
+      message: metric(totalStars),
+      color: 'blue',
+      link: [`https://github.com/${user}`],
+    }
+  }
+
+  async fetch({ user, org, affiliations, nextCursor }) {
+    const query = org === undefined ? userQuery : orgQuery
+    const variables = { user, affiliations, nextCursor }
+    return await this._requestGraphql({
+      query,
+      variables,
+      schema,
+      transformErrors: e =>
+        transformErrors(e, org === undefined ? 'user' : 'org'),
+    })
+  }
+
+  transform(repos) {
+    const totalStars = repos
+      .map(element => element.stargazers.totalCount)
+      .reduce((accumulator, currentValue) => accumulator + currentValue, 0)
+    return {
+      totalStars,
+    }
+  }
+
+  async getTotalStars({ user }, { org, affiliations }) {
+    let grandTotalStars = 0
+    let fetchedReposCount = 0
+    let nextCursor = null
+    let hasNext
+    const dataType = org === undefined ? 'user' : 'organization'
+
+    do {
+      const { data } = await this.fetch({
+        user,
+        org,
+        affiliations: affiliations.split(','),
+        nextCursor,
+      })
+      const {
+        repositories: {
+          pageInfo: { hasNextPage, endCursor },
+          nodes: repos,
+        },
+      } = data[dataType]
+      hasNext = hasNextPage
+      nextCursor = endCursor
+
+      const { totalStars } = this.transform(repos)
+      grandTotalStars += totalStars
+      fetchedReposCount += repos.length
+    } while (hasNext && fetchedReposCount < MAX_REPO_LIMIT)
```

Comment:
Can we optimise this a bit. If we have already found the first repo with zero stars while processing the first page we don't need to make the second request at all - we already know all the repos on page 2 will have 0 stars.